Sources/ProjectDescription/Tasks/Tasks.swift

```diff
@@ -0,0 +1,36 @@
+import Foundation
+
+public struct Tasks: Codable, ExpressibleByArrayLiteral {
```

Comment:
After sleeping over this, I propose something different. Let me expose the rationale first.
I'd discourage the usage of `ProjectDescriptionHelpers` from the Task swift files to minimize the time it takes to load the manifest. As a user, I expect the CLI to be fast and having to compile the helpers to load the tasks will lead to a bad user experience. That leads to keeping all the tasks in a single file. We know from Fastlane that something like that leads to huge and complex automation files. The solution?
I think we can have ` Tuist/Tasks` directory where each task is represented by a swift file:

```
Tuist/
  Tasks/
    Release.swift
    BuildDebug.swift
    ExportApp.swift
```

The name of the command would be automatically inferred from the name of the task. Note that I'm exposing them directly at the root of the CLI. This is how Rails does it and developers like it:

```bash
tuist release
tuist build-debug
tuist export-app
```

Listing those tasks and exposing them as part of the CLI is fast. However, parsing the content of those files to register the arguments and the description is not instance because we have to resort to the compiler. We can leverage caching, but with a cold cache it'll take time, which it's not an ideal developer experience. Here are some alternatives:

- We can parse with [swift-syntax](https://github.com/apple/swift-syntax) instead. Since we are parsing and not compiling, the performance should be much better.
- Come up with a convention of comments that developers can use to annotate the task. For example:

```swift
// BuildDebug.swift
// Description: Build the app for debug
// Argument: ....

let task = Task(....
```

Reply:
I really like the structure in multiple files 👏  I agree that the current proposal would lead to a large file as `Fastfile`s often do.

> I'd discourage the usage of ProjectDescriptionHelpers from the Task swift files to minimize the time it takes to load the manifest

Yes, compiling `ProjectDescriptionHelpers` will result in some performance hit, however, I believe we should allow users to share code between their templates, so something like that will be necessary (this would not have been necessary with a single file-structure but it is with multiple files). We can introduce a new framework `ProjectAutomationHelpers` to ensure that only the code that is needed will be compiled (in `ProjectDescriptionHelpers` there will be likely code not tied to tasks). 

> Note that I'm exposing them directly at the root of the CLI. This is how Rails does it and developers like it:

This should be possible to do but it means that we should make sure the loading is instant because loading all the dynamic arguments will have to be done every time *any* tuist command is invoked - not only `tuist task xxx`. So, if we decide to go down this path, the loading of commands and arguments should be as close to instantaneous as possible.

> However, parsing the content of those files to register the arguments and the description is not instance because we have to resort to the compiler. We can leverage caching, but with a cold cache it'll take time, which it's not an ideal developer experience. Here are some alternatives:

As mentioned, we should make compile the arguments without resorting to compiler if we make the task commands part of the root `tuist command`

> We can parse with swift-syntax instead. Since we are parsing and not compiling, the performance should be much better.

I can try to look into this and see how fast it is and how hard it will be to implement - I have never used `swift-syntax`, though, so might take me some time 😅 (but it should be possible)

> Come up with a convention of comments that developers can use to annotate the task. For example:

I think that this goes a little bit against what users are used to and why a lot of them use tuist - type-safety. If `swift-syntax` is not feasible or slow, rather than doing this, I'd take the performance hit (but we would have to hide behind `task` command, so the performance hit is not for all tuist commands)

Reply:
Really like the multiple files for tasks, where the name of the file is the task name. 

I do think we can get to a point where its a little confusing if we start adding more helper modules but if this would drastically speed up the time it takes to load & execute the task then its a tradeoff we can consider.

I agree with @fortmarek that comments for annotation don't fit well with the approach the project currently has, It's not type-safe and can be difficult to teach as the compiler cannot help us here, but I see how it allows us to not have to compile the file before running it. However, this concept of compiling the manifests seems to be a big selling point for Tuist and I wonder if users are okay with this trade off.

Reply:
I agree with you in regards to the type-safety being one of the reasons why people choose Tuist. My inclination is towards giving swift-syntax a shot. If we manage to get almost-instant parsing without compromising the type-safety I'd go for that. 
By the way, I've been playing with Rust a lot lately, and Cargo has the [concept of binaries](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#binaries) that serve the same purpose as what we are proposing here.